File: Structural_Patterns/Bridge/cpp/Abstraction.cpp
/*!
 *  \file       Abstraction.cpp
 *  \brief      
 *  
 */


#include "Abstraction.hpp"
#include "Implementor.hpp"

#include "UnimplementedException.hpp"

Abstraction::Abstraction(Implementor *implementor) : m_implementor{implementor}
{
}

Abstraction::~Abstraction()
{
}

std::string Abstraction::Operation() const
{
    return std::string("Abstraction: Base operation with: \n") + m_implementor->OperationImpl();
}

void Abstraction::check(int num) const volatile
{
    (void)num;
    NOT_IMPLEMENTED_EXCEPTION();
    // UNIMPLEMENTED_EXCEPTION("Abstraction");
}

File: Structural_Patterns/Bridge/c#/Abstraction.cs
/*!
 *  \file       Abstraction.cs
 *  \brief      
 *  
 */


using System;

using DesignPatterns.Structural_Patterns.Bridge.IImplementation;


namespace DesignPatterns.Structural_Patterns.Bridge
{
    class Abstraction
    {
        protected IImplementation m_implementor;

        public Abstraction(IImplementation implementor)
        {
            this.m_implementor = implementor;
        }

        public virtual string Operation()
        {
            return "Abstraction: Base operation with: \n" + m_implementor.OperationImpl();
        }
    }
}